appcues/src/main/java/com/appcues/analytics/Analytics.kt

```diff
@@ -0,0 +1,188 @@
+import com.appcues.analytics.AnalyticsIntent.UpdateGroup
+import com.appcues.analytics.AnalyticsIntent.UpdateProfile
+import com.appcues.analytics.AnalyticsIntentQueue.IntentProcessor
+import com.appcues.data.model.Experience
+import com.appcues.data.model.ExperienceTrigger.Qualification
+import com.appcues.data.model.QualificationResult
+import com.appcues.monitor.ApplicationMonitor
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.SharedFlow
+import kotlinx.coroutines.launch
+import java.util.Date
+
+internal interface QualificationService {
+
+    // will merge and hit the backend to see if it qualifies, returning QualificationResult
+    suspend fun qualify(intents: List<AnalyticsIntent>): QualificationResult?
+}
+
+internal interface RenderingService {
+
+    // will attempt to show experiences by order
+    // clearCache is mostly used when we know the qualification is from a screen_view
+    suspend fun show(experiences: List<Experience>, clearCache: Boolean)
+
+    suspend fun reset()
+}
+
+internal interface SessionService {
+
+    // returns whether session is started or not
+    suspend fun isSessionStarted(): Boolean
+
+    // starts session (can start using incoming intent or existing stored information
+    suspend fun startSession(intent: AnalyticsIntent): Boolean
+
+    // decorates incoming intent and returns a new intent with session/auto props information
+    suspend fun decorateIntent(intent: AnalyticsIntent): AnalyticsIntent
+
+    suspend fun reset()
+}
+
+// new model for Intents available for Analytics
+internal sealed class AnalyticsIntent(open val isInternal: Boolean, val timestamp: Date = Date()) {
+
+    object Anonymous : AnalyticsIntent(false)
+
+    data class Identify(
+        val userId: String,
+        val properties: Map<String, Any>?
+    ) : AnalyticsIntent(false)
+
+    data class UpdateProfile(
+        val properties: Map<String, Any>?,
+        override val isInternal: Boolean
+    ) : AnalyticsIntent(isInternal)
+
+    data class UpdateGroup(
+        val groupId: String?,
+        val properties: Map<String, Any>?,
+        override val isInternal: Boolean
+    ) : AnalyticsIntent(isInternal)
+
+    data class Event(
+        val name: String,
+        val attributes: Map<String, Any>?,
+        val context: Map<String, Any>?,
+        override val isInternal: Boolean
+    ) : AnalyticsIntent(isInternal)
+}
+
+internal class Analytics(
+    private val appcuesCoroutineScope: AppcuesCoroutineScope,
+    private val intentQueue: AnalyticsIntentQueue,
+    private val qualificationService: QualificationService,
+    private val renderingService: RenderingService,
+    private val sessionService: SessionService,
+) : IntentProcessor, ApplicationMonitor.Listener {
+
+    companion object {
+
+        const val EVENT_ATTR_SCREEN_TITLE = "screenTitle"
+        const val EVENT_CONTEXT_SCREEN_TITLE = "screen_title"
+    }
+
+    init {
+        intentQueue.setProcessor(this)
```

Comment:
following the flow: this class pushes things into the queue, then the queue calls back here to `process` things as needed based on the logic around batching items.

There is still a circular dependency there it seems, and I am wondering if there is any situation you could get in where `process` ends up spawning multiple coroutines in parallel and not guaranteed to process the queue in order?

would it make sense to have the intentQueue actually be the one managing the coroutine and ensuring things are processed in order? maybe it could pipe things through it's own Channel even to assist (how we did it with ActionProcessor for example)

Reply:
I think this somewhat better than just a circular dependency, its a callback within a limited scope.

process could be suspend fun and we could launch coroutine scope before calling it for sure, I think we can explore this idea and ensure whenever we say Flush its going in on the same order

Reply:
yeah, key thing is that everything is sent out in the desired order, however that happens.